src/services/auto-type.service.ts

```diff
@@ -0,0 +1,936 @@
+    }
+
+    init() {
+        app.on('will-quit', () => {
+            // Unregister all shortcuts.
+            globalShortcut.unregisterAll();
+        });
+        app.on('ready', () => {
+            // Register auto-type hotkey
+            this.update();
+        });
+    }
+
+    update() {
+        globalShortcut.unregisterAll();
+        this.storageService.get<string>(ElectronConstants.enableAutoTypeKey).then((enabled) => {
+            if (enabled) {
+                this.storageService.get<string>(ElectronConstants.AutoTypeHotkeyKey).then(
+                    (accelerator) => {
+                        if (accelerator) {
+                            globalShortcut.register(accelerator, () => {
+                                this.typeTarget();
+                            });
+
+                            if (!globalShortcut.isRegistered(accelerator)) {
+                                throw new Error('Failed to register auto-type hotkey.');
+                            }
+                        }
+                    });
+            }
+        });
+    }
+
+    async getTarget(): Promise<string> {
+        return new Promise((resolve, reject) => {
+            const timer = setTimeout(reject, 10000, 'Took too long to select a window');
+            windowManager.once('window-activated', (window: Window) => {
+                clearTimeout(timer);
+                resolve(window.getTitle());
+            });
+        });
+    }
+
+    getPossibleTargets(): string[] {
+        return windowManager.getWindows().map((w: Window) => w.getTitle());
+    }
+
+    typeUsingClip(str: string) {
+        if (str.length === 0) { return; }
+        this.platformUtilsService.copyToClipboard(str);
+        robot.keyTap('v', this.systemClipModifier);
+    }
+
+    async typeTarget() {
+        const target = await this.getTarget();
+
+        // Get candidate targets
+        const candidates = [];
+        const ciphers = await this.cipherService.getAll();
+        for (const cipher of ciphers) {
+            if (cipher.enableAutoType) {
+                const decCipher = await cipher.decrypt();
+                for (const targetCandidate of decCipher.autoTypeTargets) {
+                    const match = target.match(targetCandidate.target);
+                    if (match[0].length) {
+                        candidates.push({match: match, target: targetCandidate, cipher: cipher, decCipher: decCipher});
+                    }
+                }
+            }
+        }
+
+        // Decide on final candidate
+        let candidate: {match: RegExpMatchArray, target: AutoTypeView, cipher: Cipher, decCipher: CipherView} = null;
+        if (candidates.length === 1) {
+            candidate = candidates[0];
+        } else if (candidates.length > 1) {
+            candidates.sort((a, b) => b.match.length - a.match.length);
+            candidate = candidates[0]; // TODO show context menu to pick candidate
```

Comment:
This line needs to be replaced with the candidate selection modal that still has to be written.